Keystroke propagation to previous panel/s

jmillan opened the pull request:
The motivation of this feature is being able to run an action which is not
handled (or is only partially handled) by the current panel. After
handling the key on the current panel, this is destroyed by the ui_manager
and the key is passed to the previous panel. This can be done recursively until
the last panel is reached.

A clear use case is the ability to invoke ACTION_CLEAR_CALLS action from any of
'call_list', 'call_flow', or 'call_raw' with a single keystroke.

Kaian commented:
Hi Jose!

Although it will be a _not expected behaviour_ for the common folk, I like it and I think we will find other cases where we want this. If it's properly documented in the keybindings section (p.e. leave screen and clear call list), wil make movement between panels more faster.

Sadly the header comments of key handlers function of the panels are outdated (since we changed keys for actions) and actually they never return -1. They return 0 when the key is handed in the panel, or the input keycode when not. Now that we have different return behaviours, a good enhacement will be to have defines to describe the result in ui_manager.h.

```
#define KEY_HANDLED      0
#define KEY_PROPAGATED  -1
```

For the user input loop, I will like to keep it following the same pattern (look top panel, draw it, read key, handle key), instead of making exception to this logic (and make a loop iteration without the read key part) I will do something like this.

```
    // Enable key input on current panel
    win = panel_window(panel);
    keypad(win, TRUE);

    // Get pressed key
    c = wgetch(win);
    
    // Timeout, no key pressed
    if (c == ERR)
        continue;
    
    // Handle received key
    while (c != KEY_HANDLED) {
        // Check if current panel has custom bindings for that key
        c = ui_handle_key(ui, c);

        if (c == KEY_HANDLED) {
            // Panel handled this key
            continue;
        } else if (c == KEY_PROPAGATED) {
            // Destroy current panel
            ui_destroy(ui);
            // Try to handle this key with the previus panel
            ui = ui_find_by_panel(panel_below(NULL));
        } else {
            // No panel handles this key, use default handler
            c = default_handle_key(ui, c);
        }
    }
```

An make `default_handle_key` to always return KEY_HANDLED. This way we will avoid to redraw a panel and handle a key when we actually only want to handle a key.

I'm telling this without testing anything above. What do you think?

Thanks a lot for digging in the code and making all this enhancements!!

jmillan commented:
Hi @Kaian,

>Although it will be a not expected behaviour for the common folk, I like it and I think we will find other cases where we want this. If it's properly documented in the keybindings section (p.e. leave screen and clear call list), wil make movement between panels more faster.

Agreed, I've just documented it in the help section of 'call_flow' panel.

>Sadly the header comments of key handlers function of the panels are outdated (since we changed keys for actions) and actually they never return -1.

Right, I've updated the header comments too. Key handlers return -1 if `call_flow_info` and `call_raw_info` respectively are not set. I think that confused me and made me go for the -2 :-)
 
>For the user input loop, I will like to keep it following the same pattern (look top panel, draw it, read key, handle key), instead of making exception to this logic (and make a loop iteration without the read key part) I will do something like this.

Absolutely agreed. It's been applied with minor changes. 
   
>This way we will avoid to redraw a panel and handle a key when we actually only want to handle a key.

Good point!

BTW: I've indicated the 'Ctrl-L' key in the help section even if it's not implemented in this branch, since it will be available when it's merged into master.

Thanks a lot!

Kaian commented:
Hi José,

I have added some comments to your commit. Sorry for driving you crazy here, requesting so many changes :sweat: 

Also, in case you wan't to do simple ui checks, run `make check` to be sure basic interactions haven't broken.

Thanks A LOT for your work and effort with all this PRs!!

jmillan commented:
Don't worry for the comments, I appreciate it ;-).

Kaian commented:
Awesome!

Thanks a lot for the PR! 
I'll try to update the key_handlers to avoid key-hackings asap :D

jmillan commented:
Thanks for the merge ;-)
